### PR TITLE
Add TableWriteMergeNode in Velox

### PR DESCRIPTION
**Pull request by xiaoxmeng**
This is part of multi table writer driver support: [5546](https://github.com/facebookincubator/velox/issues/5546)

### CODE REVIEW

File: velox/core/PlanNode.cpp [hunk 1]
```diff
@@ -1557,6 +1557,28 @@ PlanNodePtr TableWriteNode::create(const folly::dynamic& obj, void* context) {
       source);
 }
 
+void TableWriteMergerNode::addDetails(std::stringstream& /* stream */) const {
+  // TODO Add connector details.
+}
+
+folly::dynamic TableWriteMergerNode::serialize() const {
+  auto obj = PlanNode::serialize();
+  VELOX_CHECK_EQ(sources_.size(), 1);
```
Comment (kewang1024) [inline velox/core/PlanNode.cpp:1566]: NIT:
VELOX_CHECK_EQ(sources_.size(), 1, "TableWriteMergerNode source size must be 1");

**Review by kewang1024 — APPROVED** [review]

File: velox/core/PlanNode.h [hunk 2]
```diff
@@ -545,6 +545,37 @@ class TableWriteNode : public PlanNode {
   const connector::CommitStrategy commitStrategy_;
 };
 
+class TableWriteMergerNode : public PlanNode {
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:548]: Would be nice to update documentation at https://facebookincubator.github.io/velox/develop/operators.html

File: velox/core/PlanNode.h [hunk 3]
```diff
@@ -545,6 +545,37 @@ class TableWriteNode : public PlanNode {
...
+ public:
+  TableWriteMergerNode(
+      const PlanNodeId& id,
+      RowTypePtr outputType,
+      const PlanNodePtr& source)
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:553]: nit: pass by value, then move

File: velox/core/PlanNode.h [hunk 4]
```diff
@@ -545,6 +545,37 @@ class TableWriteNode : public PlanNode {
   const connector::CommitStrategy commitStrategy_;
 };
 
+class TableWriteMergerNode : public PlanNode {
+ public:
+  TableWriteMergerNode(
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:550]: TableWriteMergerNode -> TableWriteMergeNode (no 'r')

File: velox/core/PlanNode.h [hunk 5]
```diff
@@ -1078,7 +1109,7 @@ class PartitionedOutputNode : public PlanNode {
         replicateNullsAndAny_(replicateNullsAndAny),
         partitionFunctionSpec_(std::move(partitionFunctionSpec)),
         outputType_(std::move(outputType)) {
-    VELOX_CHECK(numPartitions > 0, "numPartitions must be greater than zero");
+    VELOX_CHECK_GT(numPartitions, 0, "numPartitions must be greater than zero");
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1112]: Thanks. The error message here doesn't add anything. Consider removing.

File: velox/exec/tests/utils/PlanBuilder.h [hunk 6]
```diff
@@ -271,6 +271,11 @@ class PlanBuilder {
       connector::CommitStrategy commitStrategy,
       const RowTypePtr outputType);
 
+  /// Add a TableWriteMergerNode.
+  ///
+  /// @param outputType The output type of table write merger node.
+  PlanBuilder& tableWriteMerger(const RowTypePtr& outputType);
```
Comment (mbasmanova) [inline velox/exec/tests/utils/PlanBuilder.h:277]: tableWriteMerger -> tableWriteMerge

File: velox/exec/tests/utils/PlanBuilder.h [hunk 7]
```diff
@@ -271,6 +271,11 @@ class PlanBuilder {
       connector::CommitStrategy commitStrategy,
       const RowTypePtr outputType);
 
+  /// Add a TableWriteMergerNode.
+  ///
+  /// @param outputType The output type of table write merger node.
```
Comment (mbasmanova) [inline velox/exec/tests/utils/PlanBuilder.h:276]: What is the output of this plan node?

File: velox/core/PlanNode.cpp [hunk 8]
```diff
@@ -1557,6 +1557,29 @@ PlanNodePtr TableWriteNode::create(const folly::dynamic& obj, void* context) {
       source);
 }
 
+void TableWriteMergeNode::addDetails(std::stringstream& /* stream */) const {
+  // TODO Add connector details.
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:1561]: Are these any connector details to add?

File: velox/core/PlanNode.h [hunk 9]
```diff
@@ -545,6 +545,42 @@ class TableWriteNode : public PlanNode {
   const connector::CommitStrategy commitStrategy_;
 };
 
+class TableWriteMergeNode : public PlanNode {
+ public:
+  /// 'outputType' specifies the type to store the metadata of table write
+  /// output which contains the following columns: 'numWrittenRows', 'fragment'
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:551]: Is the outputType parameter needed? Isn't it always the same?

File: velox/exec/tests/utils/PlanBuilder.h [hunk 10]
```diff
@@ -271,6 +271,13 @@ class PlanBuilder {
       connector::CommitStrategy commitStrategy,
       const RowTypePtr outputType);
 
+  /// Add a TableWriteMergeNode.
+  ///
+  /// @param outputType The output type of table write merge node which consists
```
Comment (mbasmanova) [inline velox/exec/tests/utils/PlanBuilder.h:276]: Can we remove this parameter? It seems like there is not much freedom in what can be specified here.

**Review by mbasmanova — APPROVED** [review]

**Pull request closed without merging**